include/libdnf/base/goal.hpp

```diff
@@ -54,53 +54,235 @@ class Goal {
-    /// Prevent reinstallation by adding of already installed packages with the same NEVRA
+
+    /// Add install job to the goal for future resolvemnt. The operation will not result in package reinstall when a
+    /// requested package is already installed. In default it uses `clean_requirements_on_remove` with `false` value,
+    /// but it can be overriden by setting.
+    ///
+    /// @param spec      A string describing the Goal install job.
+    /// @param settings  A struction to override default goal settings. The value `from_repo_ids` is not used.
     void add_rpm_install(const std::string & spec, const libdnf::GoalJobSettings & settings);
-    /// Prevent reinstallation by adding of already installed packages with the same NEVRA
+
+    /// Add install job to the goal. The operation will not result in package reinstall when requested package
+    /// with the same NEVRA is already installed. In default it uses `clean_requirements_on_remove` with `false` value.
+    ///
+    /// @param rpm_package      A package to be added to the install job.
     void add_rpm_install(const libdnf::rpm::Package & rpm_package);
-    /// Prevent reinstallation by adding of already installed packages with the same NEVRA
+
+    /// Add install job to the goal. The operation will not result in package reinstall when requested package
+    /// with the same NEVRA is already installed. In default it uses `clean_requirements_on_remove` with `false` value.
+    ///
+    /// @param rpm_package      A package to be added to the install job.
+    /// @param settings         A struction to override default goal settings. Only `strict`, `best`, and `clean_requirements_on_remove` are used.
     void add_rpm_install(const libdnf::rpm::Package & rpm_package, const libdnf::GoalJobSettings & settings);
-    /// Prevent reinstallation by adding of already installed packages with the same NEVRA
+
+    /// Add install job to the goal. The operation will not result in package reinstall when requested package
+    /// with the same NEVRA is already installed. In default it uses `clean_requirements_on_remove` with `false` value.
+    ///
+    /// @param package_set      A package_set to be added to the install job.
     void add_rpm_install(const libdnf::rpm::PackageSet & package_set);
-    /// Prevent reinstallation by adding of already installed packages with the same NEVRA
+
+    /// Add install job to the goal. The operation will not result in package reinstall when requested package
+    /// with the same NEVRA is already installed. In default it uses `clean_requirements_on_remove` with `false` value.
+    ///
+    /// @param package_set      A package_set to be added to the install job.
+    /// @param settings         A struction to override default goal settings. Only `strict`, `best`, and `clean_requirements_on_remove` are used.
     void add_rpm_install(const libdnf::rpm::PackageSet & package_set, const libdnf::GoalJobSettings & settings);
```

Comment:
Is there any chance at all to reduce the number of overloads for the add_* methods? There are six(!) overloads for each operation, it's really ugly and the docstring needs to be duplicated six times too, with slight variations...

Reply:
The methods with std::string are completely different from methods with Package or PackageSet. As you can see they replace even different functions in dnf and libdnf.

Reply:
In theory we can drop a version with Package, but this not an intention of PR. In next PR I will provide code changes.

Reply:
It is quite common to have one documentation for all overloads - may be we could pick one (or two if you consider std::string/Package/PackageSet that different) which would contain the shared part of the add_rpm_install() functionality description and the others could reference this and only describe differences (like the way how the packages are selected and whether the settings overloads are present).